issue-557 added automatically span creation for DispatcherTypes: FORWARD, INCLUDE, ERROR

kananindzya opened the pull request:
closes #557

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1212/detail/PR-1212/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1212/detail/PR-1212/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1212/detail/PR-1212/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1212/detail/PR-1212/2//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #1212 updated]`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-06-01T20:01:55.554+0000

* Duration: 3 min 4 sec


* Commit: e33b14d3c9653ded790ad9609470b83f792a204c




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-06-01T20:04:56.376+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-java/pipelines/apm-agent-java-mbp/pipelines/PR-1212/runs/2/steps/81/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-06-01T20:04:32.115Z]  > git checkout -f e33b14d3c9653ded790ad9609470b83f792a204c # timeout=15
[2020-06-01T20:04:32.340Z]  > git remote # timeout=10
[2020-06-01T20:04:32.347Z]  > git config --get remote.origin.url # timeout=10
[2020-06-01T20:04:32.362Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-01T20:04:32.368Z]  > git merge 4067420042466fd20ef704f82d6535c85fd09caa # timeout=10
[2020-06-01T20:04:32.378Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-06-01T20:04:32.394Z]  > git config core.sparsecheckout # timeout=10
[2020-06-01T20:04:32.403Z]  > git checkout -f e33b14d3c9653ded790ad9609470b83f792a204c # timeout=15
[2020-06-01T20:04:36.328Z] Commit message: "issue-557 added automatically span creation for DispatcherTypes: FORWARD, INCLUDE, ERROR"
[2020-06-01T20:04:36.347Z] First time build. Skipping changelog.
[2020-06-01T20:04:36.347Z] Cleaning workspace
[2020-06-01T20:04:36.326Z]  > git rev-list --no-walk 9dfa4cc4888141ac8d58a0930fd5be4e80864786 # timeout=10
[2020-06-01T20:04:36.344Z]  > git rev-parse --verify HEAD # timeout=10
[2020-06-01T20:04:36.361Z] Resetting working tree
[2020-06-01T20:04:36.362Z]  > git reset --hard # timeout=10
[2020-06-01T20:04:36.610Z]  > git clean -fdx # timeout=10
[2020-06-01T20:04:37.455Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2020-06-01T20:04:37.497Z] Timeout set to expire in 1 hr 0 min
[2020-06-01T20:04:37.510Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2020-06-01T20:04:37.865Z] [INFO] Number of builds to be searched 10
[2020-06-01T20:04:38.434Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-06-01T20:04:38.471Z] Running in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1212/src/github.com/elastic/apm-agent-java
[2020-06-01T20:04:38.489Z] [INFO] gitCheckout: Checkout SCM PR-1212 with some customisation.
[2020-06-01T20:04:38.517Z] [INFO] Override default checkout
[2020-06-01T20:04:38.549Z] Sleeping for 10 sec
[2020-06-01T20:04:48.708Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-06-01T20:04:48.726Z] Wiping out workspace first.
[2020-06-01T20:04:48.735Z] Cloning the remote Git repository
[2020-06-01T20:04:48.746Z] Cloning repository git@github.com:elastic/apm-agent-java.git
[2020-06-01T20:04:48.774Z]  > git init /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1212/src/github.com/elastic/apm-agent-java # timeout=10
[2020-06-01T20:04:48.785Z] Using reference repository: /var/lib/jenkins/.git-references/apm-agent-java.git
[2020-06-01T20:04:48.787Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-06-01T20:04:48.787Z]  > git --version # timeout=10
[2020-06-01T20:04:48.811Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-01T20:04:48.815Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-06-01T20:04:49.815Z] Cleaning workspace
[2020-06-01T20:04:49.837Z] Pruning obsolete local branches
[2020-06-01T20:04:49.791Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-01T20:04:49.794Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-06-01T20:04:49.803Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-06-01T20:04:49.811Z]  > git rev-parse --verify HEAD # timeout=10
[2020-06-01T20:04:49.818Z] No valid HEAD. Skipping the resetting
[2020-06-01T20:04:49.819Z]  > git clean -fdx # timeout=10
[2020-06-01T20:04:49.834Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-06-01T20:04:49.834Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-01T20:04:49.839Z]  > git fetch --tags --progress --prune -- git@github.com:elastic/apm-agent-java.git +refs/pull/1212/head:refs/remotes/origin/PR-1212 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2020-06-01T20:04:50.630Z] Merging remotes/origin/master commit 4067420042466fd20ef704f82d6535c85fd09caa into PR head commit e33b14d3c9653ded790ad9609470b83f792a204c
[2020-06-01T20:04:50.841Z] Merge succeeded, producing e33b14d3c9653ded790ad9609470b83f792a204c
[2020-06-01T20:04:50.841Z] Checking out Revision e33b14d3c9653ded790ad9609470b83f792a204c (PR-1212)
[2020-06-01T20:04:51.119Z] Commit message: "issue-557 added automatically span creation for DispatcherTypes: FORWARD, INCLUDE, ERROR"
[2020-06-01T20:04:51.119Z] Cleaning workspace
[2020-06-01T20:04:50.627Z]  > git config core.sparsecheckout # timeout=10
[2020-06-01T20:04:50.632Z]  > git checkout -f e33b14d3c9653ded790ad9609470b83f792a204c # timeout=15
[2020-06-01T20:04:50.806Z]  > git remote # timeout=10
[2020-06-01T20:04:50.811Z]  > git config --get remote.origin.url # timeout=10
[2020-06-01T20:04:50.814Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-06-01T20:04:50.819Z]  > git merge 4067420042466fd20ef704f82d6535c85fd09caa # timeout=10
[2020-06-01T20:04:50.829Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-06-01T20:04:50.838Z]  > git config core.sparsecheckout # timeout=10
[2020-06-01T20:04:50.842Z]  > git checkout -f e33b14d3c9653ded790ad9609470b83f792a204c # timeout=15
[2020-06-01T20:04:51.116Z]  > git rev-parse --verify HEAD # timeout=10
[2020-06-01T20:04:51.120Z] Resetting working tree
[2020-06-01T20:04:51.120Z]  > git reset --hard # timeout=10
[2020-06-01T20:04:51.142Z]  > git clean -fdx # timeout=10
[2020-06-01T20:04:51.877Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-06-01T20:04:52.506Z] + git fetch https://****:****@github.com/elastic/apm-agent-java.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-06-01T20:04:53.118Z] Archiving artifacts
[2020-06-01T20:04:54.114Z] + git rev-parse HEAD
[2020-06-01T20:04:54.469Z] + git rev-parse HEAD
[2020-06-01T20:04:54.770Z] + git rev-parse origin/pr/1212
[2020-06-01T20:04:54.813Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-06-01T20:04:55.278Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-06-01T20:04:56.119Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-java/pulls/1212/reviews return 0 elements
[2020-06-01T20:04:56.150Z] [INFO] githubPrCheckApproved: Title: issue-557 added automatically span creation for DispatcherTypes: FORWARD, INCLUDE, ERROR - User: kananindzya - Author Association: CONTRIBUTOR
[2020-06-01T20:04:56.378Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-06-01T20:04:56.378Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-06-01T20:04:56.407Z] [INFO] Let's stop build #2. The PR is not allowed to run in the CI yet
[2020-06-01T20:04:56.422Z] Sleeping for 5 sec
[2020-06-01T20:04:57.432Z] Stage "Build" skipped due to earlier failure(s)
[2020-06-01T20:04:57.524Z] Stage "Tests" skipped due to earlier failure(s)
[2020-06-01T20:04:57.586Z] Stage "Unit Tests" skipped due to earlier failure(s)
[2020-06-01T20:04:57.588Z] Stage "Smoke Tests 01" skipped due to earlier failure(s)
[2020-06-01T20:04:57.589Z] Stage "Smoke Tests 02" skipped due to earlier failure(s)
[2020-06-01T20:04:57.591Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-06-01T20:04:57.592Z] Stage "Javadoc" skipped due to earlier failure(s)
[2020-06-01T20:04:57.662Z] Failed in branch Unit Tests
[2020-06-01T20:04:57.664Z] Failed in branch Smoke Tests 01
[2020-06-01T20:04:57.665Z] Failed in branch Smoke Tests 02
[2020-06-01T20:04:57.666Z] Failed in branch Benchmarks
[2020-06-01T20:04:57.667Z] Failed in branch Javadoc
[2020-06-01T20:04:57.727Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-06-01T20:04:57.774Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-01T20:04:57.794Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-01T20:04:57.831Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-06-01T20:04:58.347Z] Running on Jenkins in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1212@2
[2020-06-01T20:04:58.626Z] [INFO] getVaultSecret: Getting secrets
[2020-06-01T20:04:58.682Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-06-01T20:04:59.340Z] + chmod 755 generate-build-data.sh
[2020-06-01T20:04:59.340Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1212/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1212/runs/2 ABORTED 183527
[2020-06-01T20:05:00.251Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1212/runs/2/steps/?limit=10000 -o steps-info.json

```

</p>
</details>

kananindzya commented:
```
ebsphere] 2020-06-02 05:53:49,984 [elastic-apm-circuit-breaker] INFO  co.elastic.apm.agent.impl.circuitbreaker.CircuitBreaker - Stopping the Circuit Breaker thread.
[websphere] 2020-06-02 05:53:49,984 [Framework stop] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Type match for instrumentation ExecutorRunnableInstrumentation: ((((hasSuperType(erasure(name(equals(java.util.concurrent.Executor)))) and not(name(equals(org.apache.felix.resolver.ResolverImpl$DumbExecutor)))) and not(name(startsWith(com.hazelcast)))) and not((name(contains($Proxy)) or name(contains($$))))) and not(isInterface())) matches class java.util.concurrent.ForkJoinPool
[websphere] 2020-06-02 05:53:49,985 [Framework stop] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Type match for instrumentation ExecutorCallableInstrumentation: ((((hasSuperType(erasure(name(equals(java.util.concurrent.Executor)))) and not(name(equals(org.apache.felix.resolver.ResolverImpl$DumbExecutor)))) and not(name(startsWith(com.hazelcast)))) and not((name(contains($Proxy)) or name(contains($$))))) and not(isInterface())) matches class java.util.concurrent.ForkJoinPool
[websphere] [AUDIT   ] CWWKT0017I: Web application removed (default_host): https://08df9efac3d5:9443/simple-webapp/
[websphere] [AUDIT   ] CWWKT0017I: Web application removed (default_host): https://08df9efac3d5:9443/cdi-app/
[websphere] [AUDIT   ] CWWKT0017I: Web application removed (default_host): https://08df9efac3d5:9443/jsf-http-get/
[websphere] 2020-06-02 05:53:50,033 [Framework stop] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Method match for instrumentation ExecutorRunnableInstrumentation: ((((name(equals(execute)) and returns(erasure(is(void)))) and hasParameter(hasTypes(erasures(containing(is(interface java.lang.Runnable)))))) or ((name(equals(submit)) and returns(erasure(is(interface java.util.concurrent.Future)))) and hasParameter(hasTypes(erasures(containing(is(interface java.lang.Runnable))))))) or ((name(equals(submit)) and returns(erasure(is(interface java.util.concurrent.Future)))) and hasParameter(hasTypes(erasures(containing(is(interface java.lang.Runnable), is(class java.lang.Object))))))) matches public void java.util.concurrent.ForkJoinPool.execute(java.lang.Runnable)
[websphere] [AUDIT   ] CWWKZ0009I: The application simple-webapp has stopped successfully.
[websphere] [AUDIT   ] CWWKZ0009I: The application cdi-app has stopped successfully.
[websphere] [AUDIT   ] CWWKZ0009I: The application jsf-http-get has stopped successfully.
[websphere] [AUDIT   ] CWWKE0036I: The server defaultServer stopped after 15.941 seconds.
Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 53.62 sec <<< FAILURE! - in co.elastic.apm.servlet.WebSphereIT
testAllScenarios[WebSphere 8.5.5](co.elastic.apm.servlet.WebSphereIT)  Time elapsed: 15.743 sec  <<< FAILURE!
org.opentest4j.AssertionFailedError: 

Expecting:
 <5>
to be equal to:
 <2>
but was not.

testAllScenarios[WebSphere webProfile7](co.elastic.apm.servlet.WebSphereIT)  Time elapsed: 9.87 sec  <<< FAILURE!
org.opentest4j.AssertionFailedError: 

Expecting:
 <3>
to be equal to:
 <2>
but was not.


Results :

Failed tests: 
  JBossIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <3>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  TomcatIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <4>
to be equal to:
 <2>
but was not.
  WebSphereIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <5>
to be equal to:
 <2>
but was not.
  WebSphereIT>AbstractServletContainerIntegrationTest.testAllScenarios:264 
Expecting:
 <3>
to be equal to:
 <2>
but was not.

Tests run: 28, Failures: 9, Errors: 0, Skipped: 1
```

kananindzya commented:
with last fix, now I have problem with `ServletApiTestApp#testTransactionReportingWithErrorHandling`
It fails only on WebsphereIT(webProfile7).
Transaction by some reason not created.

kananindzya commented:
@felixbarny, I noticed that for some Jetty application container span of DispatcherType.ERROR created, but for example websphere, widlfy, tomcat - not.

kananindzya commented:
when we call `"/simple-webapp" + "/unknown"` in`ServletApiTestApp#testTransactionReportingWithErrorHandling`
 JettyIT(9.3) fails with error:

```
ailed tests: 
  JettyIT>AbstractServletContainerIntegrationTest.testAllScenarios:264->AbstractServletContainerIntegrationTest.assertTransactionReported:273->AbstractServletContainerIntegrationTest.getAllReported:338->AbstractServletContainerIntegrationTest.getReportedTransactions:414->AbstractServletContainerIntegrationTest.getEvents:452->AbstractServletContainerIntegrationTest.validateEventMetadata:475->AbstractServletContainerIntegrationTest.validateServiceName:492 [Event has non-expected service name {"timestamp":1591161440603002,"name":"DefaultServlet#doGet","id":"7c4a6aab6a74eda3","trace_id":"191fdaf7ed5f2d2d6876d647b6402b22","type":"request","duration":18.712,"result":"HTTP 4xx","context":{"service":{"name":"Simple Web App"},"request":{"method":"GET","headers":{"Connection":"keep-alive","User-Agent":"okhttp/3.9.1","Host":"localhost:33485","Accept-Encoding":"gzip"},"url":{"full":"http://localhost:33485/simple-webapp/unknown","hostname":"localhost","port":"33485","pathname":"/simple-webapp/unknown","protocol":"http"},"socket":{"encrypted":false,"remote_address":"172.30.0.1"},"http_version":"1.1"},"response":{"finished":true,"headers_sent":true,"status_code":404},"tags":{}},"span_count":{"dropped":0,"started":1},"sampled":true}] 
Expecting:
 <"Simple Web App">
to be equal to:
 <"jsf-http-get">
but was not.
  JettyIT>AbstractServletContainerIntegrationTest.testAllScenarios:264->AbstractServletContainerIntegrationTest.assertTransactionReported:273->AbstractServletContainerIntegrationTest.getAllReported:338->AbstractServletContainerIntegrationTest.getReportedTransactions:414->AbstractServletContainerIntegrationTest.getEvents:452->AbstractServletContainerIntegrationTest.validateEventMetadata:475->AbstractServletContainerIntegrationTest.validateServiceName:492 [Event has non-expected service name {"timestamp":1591161454235002,"name":"DefaultServlet#doGet","id":"a3575d7702a7837d","trace_id":"965887bd126bd2e4e429bf74db52d090","type":"request","duration":16.391,"result":"HTTP 4xx","context":{"service":{"name":"Simple Web App"},"request":{"method":"GET","headers":{"Connection":"keep-alive","User-Agent":"okhttp/3.9.1","Host":"localhost:33486","Accept-Encoding":"gzip"},"url":{"full":"http://localhost:33486/simple-webapp/unknown","hostname":"localhost","port":"33486","pathname":"/simple-webapp/unknown","protocol":"http"},"socket":{"encrypted":false,"remote_address":"172.30.0.1"},"http_version":"1.1"},"response":{"finished":true,"headers_sent":true,"status_code":404},"tags":{}},"span_count":{"dropped":0,"started":1},"sampled":true}] 
Expecting:
 <"Simple Web App">
to be equal to:
 <"jsf-http-get">
but was not.

```

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -143,6 +156,37 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                    spanAction = INCLUDE_SPAN_ACTION;
+                    isAllowedType = true;
+                } else if (dispatcherType == DispatcherType.ERROR) {
+                    spanName = ERROR + SPACE + request.getAttribute(RequestDispatcher.FORWARD_SERVLET_PATH);
```
Why using `RequestDispatcher.FORWARD_SERVLET_PATH` in this case? I think we want the error page path. The transaction will contain the original servlet path.

kananindzya replied:
changed to getting  
```
Object servletPath = request.getServletPath();
```

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -143,6 +156,37 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                } else if (dispatcherType == DispatcherType.ERROR) {
+                    spanName = ERROR + SPACE + request.getAttribute(RequestDispatcher.FORWARD_SERVLET_PATH);
+                    spanAction = ERROR_SPAN_ACTION;
+                    isAllowedType = true;
```
Please check if you already get the Exception object stored in the `RequestDispatcher.ERROR_EXCEPTION` attribute somewhere else. If not, let's create and send an error here with `captureException` through the parent span.

eyalkoren commented on integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/tests/ServletApiTestApp.java:
```diff
@@ -75,6 +78,58 @@ private void testCaptureBody(AbstractServletContainerIntegrationTest test) throw
+        List<JsonNode> reportedSpans = test.getReportedSpans();
+        assertThat(reportedSpans.size()).isEqualTo(2);
+
+        for (JsonNode span : test.getReportedSpans()) {
+            String spanType = span.get("type").textValue();
+            if ("servlet.request-dispatcher.forward".equals(spanType)) {
+                isExistForwardSpan = true;
+                assertThat(span.get("name").textValue()).isEqualTo("FORWARD /servlet");
+            } else if ("db.h2.query".equals(spanType)) {
+                isExistDbH2QuerySpan = true;
+            }
+        }
```
You need to assert that exactly the expected spans are captured. Potentially, this can pass if two non-related spans are captured.
You can `filter` the span collection for each and assert you get exactly one that contains the required info.

kananindzya replied:
added filter before assert.

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -60,6 +63,16 @@
+    private static final String FORWARD_SPAN_ACTION = "forward";
+    private static final String INCLUDE_SPAN_ACTION = "include";
+    private static final String ERROR_SPAN_ACTION = "error";
+    private static final String FORWARD = "FORWARD";
+    private static final String INCLUDE = "INCLUDE";
+    private static final String ERROR = "ERROR";
+    private static final String SPACE = " ";
+    private static final String EMPTY = "";
```
```suggestion
    private static final String FORWARD = "FORWARD ";
    private static final String INCLUDE = "INCLUDE ";
    private static final String ERROR = "ERROR ";
```

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                servletTransactionHelper.fillRequestContext(transaction, request.getProtocol(), request.getMethod(), request.isSecure(),
+                    request.getScheme(), request.getServerName(), request.getServerPort(), request.getRequestURI(), request.getQueryString(),
+                    request.getRemoteAddr(), request.getHeader("Content-Type"));
+            } else if (transaction == null) {
```
Why is the `transaction == null` required here? Seems it will always be `null`

kananindzya replied:
i changed here condition to -> 
```
servletRequest.getDispatcherType() != DispatcherType.ASYNC
```

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                        if (pathInfo != null) {
+                            span.appendToName((String) pathInfo);
+                        }
+                        span.withAction(FORWARD_SPAN_ACTION);
```
```suggestion
                        span.withAction(INCLUDE_SPAN_ACTION);
```

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                if (parent != null) {
+                    final HttpServletRequest request = (HttpServletRequest) servletRequest;
+                    DispatcherType dispatcherType = request.getDispatcherType();
+                    boolean isSpannableDispatcherType = false;
+                    span = parent.createSpan()
+                        .withType(SPAN_TYPE)
+                        .withSubtype(SPAN_SUBTYPE);
```
Do not create a span unless you know we need one. Currently these spans can leak. You must always end a span you created, like you must always deactivate a span you activated (which you did).
In rare cases (this is not one), you may create a span and then `requestDiscarding` it, but you still must end it.

eyalkoren commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                        span.withAction(ERROR_SPAN_ACTION);
+                        isSpannableDispatcherType = true;
+                    }
+                    if (isSpannableDispatcherType && !parent.getNameAsString().equals(span.getNameAsString())) {
```
`getNameAsString` is allocating a String objects (so two in this comparison).
There are better ways to avoid creation of same nested spans. For example, before creating a span, see if the parent span has the same type, subtype and action. 
What cases are you trying to prevent with this comparison?

kananindzya commented on apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:
```diff
@@ -143,6 +156,37 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                String spanAction = null, spanName = null;
+                if (dispatcherType == DispatcherType.FORWARD) {
+                    String pathInfo = request.getPathInfo();
+                    spanName = FORWARD + SPACE + request.getServletPath() + (pathInfo != null ? pathInfo : EMPTY);
```
changed to appendToName

eyalkoren commented:
@kananindzya thanks for taking this back. What's the state of this PR? Is it ready for another review?

kananindzya commented:
> @kananindzya thanks for taking this back. What's the state of this PR? Is it ready for another review?

I updated yesterday as per the new changes. But there was a problem with integration tests, the isntrumentation for some reason not working. Today I will try to find the reason.
Next, I should add exception capture from http headers. 
And change logic with comparing parent span type, subtype, action, instead of comparing `nameAsString`.

kananindzya commented:
> Thanks @kananindzya, looking good!
> 
> I started commenting but decided it's easier to code instead slightly_smiling_face .
> Please see my [suggested changes](https://github.com/kananindzya/apm-agent-java/pull/6) and if you feel it doesn't contradict anything you intended to do, please merge.
> I mainly moved a couple of things around, saving some allocations.
> 
> I also removed a crash report you accidentally pushed but I wonder - does this need to trouble us? Do you know what caused this crash?

Great! Thanks you. I added exception handling.

kananindzya commented:
> I also removed a crash report you accidentally pushed but I wonder - does this need to trouble us? Do you know what caused this crash?

No it shouldn't trouble you. Next time I will be more attentive before pushing changes.

kananindzya commented:
> Awesome!! tada

Thanks you

purdel commented:
Is there any way to disable these spans? They are taking a toll on our Elastic storage and we do not actually need them. I tried using the `disable_instrumentations=servlet-api` config but this causes the agent to not capture any page-load transactions anymore. Using spring-mvc here.

eyalkoren commented:
Can you share numbers? How many such spans do you get per transaction?

purdel commented:
Hi @eyalkoren. We get from only a couple to more than a hundred. We had the same issue with spring-view and we were forced to stop instrumenting it. Once we did, the amount of used storage decreased to ~20GB per day, which was manageable. But now we have 140GB a day again... Mostly jsp includes.

It is worth mentioning that we have lots of traffic,

I think the quick solution would be to downgrade to release 1.17.0 which was working fine.

eyalkoren commented:
@purdel please try [this snapshot](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp/detail/PR-1448/1/artifacts) with adding `servlet-api-dispatch` to the `disable_instrumentations` config and see if it works for you. 
(I am looking into the test failures, but I don't think they are related).

purdel commented:
@eyalkoren wow, that was fast! Trying it out.

purdel commented:
@eyalkoren looks good

eyalkoren commented:
Awesome, thanks for verifying. #1448 is now merged.

purdel commented:
@eyalkoren great, thank you very much! Any idea when the next release will come?

eyalkoren commented:
I can't say exactly, but probably within a couple of weeks frame. You can register for notifications from our repo only on new releases.
